src/Compilers/CSharp/Portable/Errors/CSDiagnosticInfo.cs

```diff
@@ -42,6 +42,16 @@ internal CSDiagnosticInfo(ErrorCode code, object[] args, ImmutableArray<Symbol>
             _additionalLocations = additionalLocations.IsDefaultOrEmpty ? SpecializedCollections.EmptyReadOnlyList<Location>() : additionalLocations;
         }
 
+        private CSDiagnosticInfo(CSDiagnosticInfo original, DiagnosticSeverity severity) : base(original, severity)
+        {
+            _additionalLocations = original._additionalLocations;
+        }
+
+        internal override DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
```

Comment:
FWIW, it seems to me that dropping information about the original instance, when creating a derived instance, is a cut-and-dry bug, and it seems perfectly fine to take this change as a starting point, even if we think more refactoring is desirable. cc @jcouv @dotnet/roslyn-compiler.

Reply:
I'm okay with that as well. Fixing the deeper design issue seems orthogonal and outside the scope for this bug.

Reply:
@jcouv @RikkiGibson So is it possible to merge this as-is, and follow up on https://github.com/dotnet/roslyn/issues/66037#issuecomment-1354605898 later?

Unfortunately, I'm currently putting less time for open source due to exams (it ends 25 January).

Reply:
I'll review more carefully, and given the impact, I think it's reasonable for us to push to the PR branch to resolve any issues we identify, and merge without you needing to do anything more. Thanks, and good luck on exams 😄